### PR TITLE
Not possible to get a post type's REST API namespace if modified

### PR DESCRIPTION
**Pull request by spacedmonkey**
Trac ticket: https://core.trac.wordpress.org/ticket/53656
GB issue: https://github.com/WordPress/gutenberg/issues/33420

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/class-wp-post-type.php [hunk 1]
```diff
@@ -473,6 +482,11 @@ public function set_props( $args ) {
 			$args['show_ui'] = $args['public'];
 		}
 
+		// If not set, default rest_namespace to wp/v2 if rest_base is set.
+		if ( false === $args['rest_namespace'] && ! empty( $args['rest_base'] ) ) {
```
[the remark this answers is not in the document]
Comment (spacedmonkey) [reply src/wp-includes/class-wp-post-type.php:486]: It doesn't make sense to have a namespace, show in rest / rest base is not set.

File: src/wp-includes/rest-api.php [hunk 2]
```diff
@@ -3056,7 +3056,7 @@ function rest_get_route_for_post( $post ) {
 
 	// The only two controllers that we can detect are the Attachments and Posts controllers.
 	if ( in_array( get_class( $controller ), array( 'WP_REST_Attachments_Controller', 'WP_REST_Posts_Controller' ), true ) ) {
```
Comment (spacedmonkey) [inline src/wp-includes/rest-api.php:3058]: @TimothyBJacobs Do we need this check now?

File: src/wp-admin/edit-form-blocks.php [hunk 3]
```diff
@@ -49,20 +49,21 @@ static function( $classes ) {
...
+$rest_namespace = ! empty( $post_type_object->rest_namespace ) ? $post_type_object->rest_namespace : 'wp/v2';
 
 // Preload common data.
 $preload_paths = array(
 	'/',
 	'/wp/v2/types?context=edit',
 	'/wp/v2/taxonomies?per_page=-1&context=edit',
 	'/wp/v2/themes?status=active',
-	sprintf( '/wp/v2/%s/%s?context=edit', $rest_base, $post->ID ),
+	sprintf( '/%s/%s/%s?context=edit', $rest_namespace, $rest_base, $post->ID ),
```
Comment (TimothyBJacobs) [inline src/wp-admin/edit-form-blocks.php:61]: Let's switch this to use `rest_get_route_for_post` while we're here

File: src/wp-includes/rest-api/endpoints/class-wp-rest-autosaves-controller.php [hunk 4]
```diff
@@ -67,8 +67,8 @@ public function __construct( $parent_post_type ) {
 
 		$this->parent_controller    = $parent_controller;
 		$this->revisions_controller = new WP_REST_Revisions_Controller( $parent_post_type );
-		$this->namespace            = 'wp/v2';
 		$this->rest_base            = 'autosaves';
+		$this->namespace            = ! empty( $post_type_object->rest_namespace ) ? $post_type_object->rest_namespace : 'wp/v2';
```
[the remark this answers is not in the document]
Comment (TimothyBJacobs) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-autosaves-controller.php:71]: ( Same for other instances of the same check )

[the remark this answers is not in the document]
Comment (spacedmonkey) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-autosaves-controller.php:71]: My worry here, that there is post type object is created in a strange way and that somehow this property does not exist. Or that it is set to false for some reason. This is just a little hardnessing.

File: src/wp-admin/edit-form-blocks.php (hunk 3 above)
Comment (spacedmonkey) [reply src/wp-admin/edit-form-blocks.php:61]: Does that contain the rest_url as well?

**Comment by spacedmonkey** [comment]
> Tests look great, thanks @spacedmonkey! Just a question about the naming.

Function change name to `rest_get_route_for_post_type_items`.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php [hunk 5]
```diff
@@ -331,12 +336,18 @@ public function get_item_schema() {
...
 				),
-				'rest_base'    => array(
+				'rest_base'      => array(
 					'description' => __( 'REST base route for the post type.' ),
 					'type'        => 'string',
 					'context'     => array( 'view', 'edit', 'embed' ),
 					'readonly'    => true,
 				),
+				'rest_namespace' => array(
+					'description' => __( 'REST namespace route for the post type.' ),
```
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php:346]:
```suggestion
					'description' => __( 'REST route\'s namespace for the post type.' ),
```

**Comment by TimothyBJacobs** [comment]
So upon further review, we are going to need to remove the specific controller class checks from this method as well as the `rest_get_route_for_post` method. Otherwise, anyone who has a custom controller, even if they have registered the routes as Core expects, as we are now using these methods for beyond generating informative links.

**Comment by spacedmonkey** [comment]
> So upon further review, we are going to need to remove the specific controller class checks from this method as well as the rest_get_route_for_post method. Otherwise, anyone who has a custom controller, even if they have registered the routes as Core expects, as we are now using these methods for beyond generating informative links.

I noticed this myself. I was going to note it the dev blog after release. 

I wasn't 100% why the check was in the there in the first place, felt like over kill to me. I also don't like that you have to spin up a instance of the class to do this check. Much cleaner to just use the `show_in_rest` param.

**Comment by TimothyBJacobs** [comment]
Merged in https://core.trac.wordpress.org/changeset/51962.

**Pull request closed without merging**